### PR TITLE
[new API] add paddle.kthvalue and paddle.Tensor.kthvalue

### PR DESCRIPTION
**Pull request by zoooo0820**
### PR types
New features

### PR changes
OPs

### Describe
This pr adds API/OPs paddle.kthvalue and paddle.Tensor.kthvalue. 

kthvalue-API文档-PaddlePaddle深度学习平台 - 10 136 157 23

kthvalue-API Document-PaddlePaddle Deep Learning Platform - 10 136 157 23

### CODE REVIEW

File: paddle/fluid/operators/kthvalue_op.cu [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+    return 256;
+  else if (col > 64 && col <= 128)
+    return 128;
+  else
+    return 64;
+}
+
+template <typename T>
+bool SortKthvalue(const platform::CUDADeviceContext& ctx,
+                  const framework::Tensor* input_tensor, const int64_t num_cols,
```
Comment (sneaxiy) [inline paddle/fluid/operators/kthvalue_op.cu:45]: Too many duplicate codes with `top_k_function_cuda.h`.

File: paddle/fluid/operators/kthvalue_op.cu [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    if (axis == in_dims.size() - 1) {
+      const int64_t& input_height = framework::product(
+          framework::slice_ddim(in_dims, 0, in_dims.size() - 1));
+      const int64_t& input_width = in_dims[in_dims.size() - 1];
+      const auto& dev_ctx = ctx.cuda_device_context();
+      if (SortKthvalue<T>(dev_ctx, input, input_width, input_height, k, output,
+                          indices)) {
+        return;
+      } else {
+        LOG(INFO) << "KthvalueOP: Some errors happened when use cub sorting";
```
Comment (sneaxiy) [inline paddle/fluid/operators/kthvalue_op.cu]: Throw exception if there is any error. Do not just logging.

File: paddle/fluid/operators/kthvalue_op.cu [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+            ndims, dev_ctx, trans_ind, indices, trans);
+        TransCompute<platform::CUDADeviceContext, T>(ndims, dev_ctx, trans_out,
+                                                     output, trans);
+        if (!keepdim) {
+          output->Resize(out_dims);
+          indices->Resize(out_dims);
+        }
+        return;
+      } else {
+        LOG(INFO) << "KthvalueOP: Some errors happened when use cub sorting";
```
Comment (sneaxiy) [inline paddle/fluid/operators/kthvalue_op.cu]: Same above.

File: paddle/fluid/operators/kthvalue_op.cc [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    OP_INOUT_CHECK(ctx->HasInput("X"), "Input", "X", "kthvalue");
+    OP_INOUT_CHECK(ctx->HasOutput("Out"), "Output", "Out", "kthvalue");
+    OP_INOUT_CHECK(ctx->HasOutput("Indices"), "Output", "Indices", "kthvalue");
+    auto input_dims = ctx->GetInputDim("X");
+    const int& dim_size = input_dims.size();
+    int axis = static_cast<int>(ctx->Attrs().Get<int>("axis"));
+    PADDLE_ENFORCE_EQ(
+        (axis < dim_size) && (axis >= (-1 * dim_size)), true,
```
Comment (sneaxiy) [inline paddle/fluid/operators/kthvalue_op.cc]: Try to write:

`PADDLE_ENFORCE_LT(axis, dim_size)`

and 

`PADDLE_ENFORCE_GE(axis, -dim_size)`

File: paddle/fluid/operators/kthvalue_op.cc [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    framework::DDim dims = framework::make_ddim(dimvec);
+    ctx->SetOutputDim("Out", dims);
+    ctx->SetOutputDim("Indices", dims);
+    ctx->ShareLoD("X", "Out");
+    ctx->ShareLoD("X", "Indices");
+  }
+
+ protected:
+  framework::OpKernelType GetExpectedKernelType(
+      const framework::ExecutionContext& ctx) const override {
```
Comment (sneaxiy) [inline paddle/fluid/operators/kthvalue_op.cc:78]: Just write:
`return framework::OpKernelType(OperatorWithKernel::IndicateVarDataType(ctx, "X"), ctx.GetPlace())`

File: paddle/fluid/operators/kthvalue_op.h [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+      output_data[i * input_width + e_indices(i, 0)] = e_input(i, 0);
+    }
+  }
+}
+
+template <typename DeviceContext, typename T>
+class KthvalueCPUKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    auto* input = context.Input<framework::Tensor>("X");
```
Comment (sneaxiy) [inline paddle/fluid/operators/kthvalue_op.h:95]: Too many duplicate codes in CPU/CUDA kernels.

File: paddle/fluid/operators/kthvalue_op.cc (hunk 4 above)
Comment (zoooo0820) [reply paddle/fluid/operators/kthvalue_op.cc]: done

File: paddle/fluid/operators/kthvalue_op.cc (hunk 5 above)
Comment (zoooo0820) [reply paddle/fluid/operators/kthvalue_op.cc:78]: done

**Review by jeff41404 — DISMISSED** [review]
LGTM

**Review by sneaxiy — APPROVED** [review]
LGTM. 

TODO: do code clean to avoid too many duplicate codes. @zoooo0820

**Review by dingjiaweiww — APPROVED** [review]

**Review by jeff41404 — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LG API

**Pull request merged**